pymedphys/_dicom/anonymise/__init__.py

```diff
@@ -735,17 +903,90 @@ def get_anonymous_replacement_value(
             keyword,
             vr,
             current_value,
+            error_info,
         )
-        raise
+        raise KeyError
 
+    if keyword == "SOPInstanceUID":
+        logging.debug("Returning SOP Instance UID: %s", replacement_value)
     return replacement_value
+
+
+def is_valid_strategy_for_keywords(
+    identifying_keywords=None, replacement_strategy=None
+):
+    if identifying_keywords is None:
+        identifying_keywords = get_default_identifying_keywords()
+
+    if replacement_strategy is None:
+        replacement_strategy = get_copy_of_strategy()
+    else:
+        if (
+            "replace_values" in replacement_strategy
+            and not replacement_strategy["replace_values"]
+        ):
+            return False
+    # replace_values should always be true within a strategy.
+    # if the user wants to clear values, the underlying dispatch/strategy for that is
+    # strategy.ANONYMISATION_CLEARVALUES_DISPATCH
+    # which should be obtained using
+    # get_clearing_strategy()
+
+    baseline_keyword_vr_dict = get_baseline_keyword_vr_dict()
+    for keyword in identifying_keywords:
+        vr = baseline_keyword_vr_dict[keyword]
+        # pydicom.datadict.dictionary_VR(pydicom.datadict.tag_for_keyword(keyword))
+        if vr not in replacement_strategy:
+            return False
+    return True
+
+
+@functools.lru_cache()
+def get_copy_of_strategy():
+
+    strategy_map = immutables.Map(strategy.ANONYMISATION_HARDCODE_DISPATCH)
+    with strategy_map.mutate() as strategy_copy:
+        strategy_copy["strategy_name"] = "hardcode"
+        strategy_copy["replace_values"] = True
+        strategy_copy["keywords_to_leave_unchanged"] = ()
+        strategy_copy["delete_private_tags"] = True
+        strategy_copy["delete_unknown_tags"] = None
+        strategy_copy["copy_dataset"] = True
+        strategy_copy["identifying_keywords"] = get_default_identifying_keywords()
+        strategy_map = strategy_copy.finish()
+    is_valid_strategy_for_keywords(
+        strategy_map["identifying_keywords"], replacement_strategy=strategy_map
+    )
+    return strategy_map
+
+
+@functools.lru_cache()
+def get_clearing_strategy():
+
+    strategy_map = immutables.Map(strategy.ANONYMISATION_CLEARVALUES_DISPATCH)
+    with strategy_map.mutate() as strategy_copy:
+        strategy_copy["strategy_name"] = "clearing"
+        strategy_copy["replace_values"] = True
+        strategy_copy["keywords_to_leave_unchanged"] = ()
+        strategy_copy["delete_private_tags"] = True
+        strategy_copy["delete_unknown_tags"] = None
+        strategy_copy["copy_dataset"] = True
+        strategy_copy["identifying_keywords"] = get_default_identifying_keywords()
+        strategy_map = strategy_copy.finish()
+    is_valid_strategy_for_keywords(
+        strategy_map["identifying_keywords"], replacement_strategy=strategy_map
+    )
+    return strategy_map
```

Comment:
Given the size of this file, it's getting a unwieldy. We need to split out key concepts into their own file. It would make sense that these functions here would be split over about three files I would imagine.

It would make sense to me to have the top-level user exposed facade functions all in one file called `api.py`, any strategy based functions be within a `strategy.py`, and then maybe make a `core.py` for the rest of the code...

Reply:
Agreed.